File: src/ui_gtk/pages/explore.py
"""
ui_gtk/pages/explore.py — Explore/Browse page for GTK frontend.

Shows top-level category grid → subcategory list → app grid.
"""
import os
import threading
import gi
gi.require_version("Adw", "1")
gi.require_version("Gtk", "4.0")
from gi.repository import Adw, Gtk, GLib, Pango

from backend import packages as _pkg
from ..icon_loader import resolve_icon_path

CATEGORY_TREE = [
    ("Games",        "Game",        "applications-games-symbolic", [
        ("Action",       "ActionGame"),   ("Adventure",    "AdventureGame"),
        ("Arcade",       "ArcadeGame"),   ("Board Games",  "BoardGame"),
        ("Emulators",    "Emulator"),     ("Role Playing", "RolePlaying"),
        ("Shooter",      "Shooter"),      ("Strategy",     "StrategyGame"),
    ]),
    ("Internet",     "Network",     "applications-internet-symbolic", [
        ("Web Browsers", "WebBrowser"),   ("Email",        "Email"),
        ("Chat &amp; IM",    "InstantMessaging"), ("File Sharing","FileTransfer"),
        ("News &amp; RSS",   "News"),
    ]),
    ("Audio &amp; Video","AudioVideo",  "applications-multimedia-symbolic", [
        ("Music",        "Music"),        ("Video",        "Video"),
        ("Editors",      "AudioVideoEditing"), ("Podcasts","Podcasting"),
    ]),
    ("Graphics",     "Graphics",    "applications-graphics-symbolic", [
        ("2D Editors",   "2DGraphics"),   ("3D Editors",   "3DGraphics"),
        ("Photography",  "Photography"),  ("Viewers",      "Viewer"),
    ]),
    ("Productivity", "Office",      "applications-office-symbolic", [
        ("Office Suite", "WordProcessor"),("Spreadsheets", "Spreadsheet"),
        ("Presentation", "Presentation"), ("Calendar",     "Calendar"),
        ("Notes",        "Notes"),
    ]),
    ("Development",  "Development", "applications-development-symbolic", [
        ("IDEs",         "IDE"),          ("Version Control","RevisionControl"),
        ("Debuggers",    "Debugger"),     ("Web Dev",      "WebDevelopment"),
    ]),
    ("System",       "System",      "applications-system-symbolic", [
        ("File Managers","FileManager"),  ("Terminal",     "TerminalEmulator"),
        ("Monitors",     "Monitor"),      ("Security",     "Security"),
    ]),
    ("Utilities",    "Utility",     "applications-utilities-symbolic", [
        ("Archiving",    "Archiving"),    ("Calculators",  "Calculator"),
        ("Clocks",       "Clock"),        ("Text Editors", "TextEditor"),
    ]),
    ("Education",    "Education",   "applications-science-symbolic", [
        ("Languages",    "Languages"),    ("Mathematics",  "Math"),
        ("Science",      "Science"),
    ]),
]


class ExplorePage(Gtk.Box):
    def __init__(self, window):
        super().__init__(orientation=Gtk.Orientation.VERTICAL)
        self._win = window
        self._set_vexpand(True)

        # Stack: "categories" | "subcategories" | "apps"
        self._nav_stack = Gtk.Stack()
        self._nav_stack.set_vexpand(True)
        self._nav_stack.set_transition_type(Gtk.StackTransitionType.SLIDE_LEFT_RIGHT)
        self.append(self._nav_stack)

        self._build_category_view()

    def _set_vexpand(self, v):
        self.set_vexpand(v)

    # ── Top-level categories ──────────────────────────────────────────────────

    def _build_category_view(self):
        scroll = Gtk.ScrolledWindow()
        scroll.set_vexpand(True)
        scroll.set_policy(Gtk.PolicyType.NEVER, Gtk.PolicyType.AUTOMATIC)

        box = Gtk.Box(orientation=Gtk.Orientation.VERTICAL, spacing=0)
        box.set_margin_top(12)
        box.set_margin_bottom(12)

        list_box = Gtk.ListBox()
        list_box.set_selection_mode(Gtk.SelectionMode.NONE)
        list_box.add_css_class("boxed-list")
        list_box.set_margin_start(12)
        list_box.set_margin_end(12)

        for label, cat, icon, subcats in CATEGORY_TREE:
            row = Adw.ActionRow()
            row.set_title(label)
            row.set_activatable(True)
            row.add_prefix(Gtk.Image.new_from_icon_name(icon))
            row.add_suffix(Gtk.Image.new_from_icon_name("go-next-symbolic"))
            row.connect("activated",
                        lambda r, l=label, c=cat, s=subcats: self.browse_category(c, l, s))
            list_box.append(row)

        box.append(list_box)
        scroll.set_child(box)
        self._nav_stack.add_named(scroll, "categories")

    # ── Subcategories ─────────────────────────────────────────────────────────

    def browse_category(self, cat: str, label: str, subcats: list = None):
        """Show subcategory list for a top-level category."""
        # Remove old subcat page if exists
        old = self._nav_stack.get_child_by_name("subcategories")
        if old:
            self._nav_stack.remove(old)

        scroll = Gtk.ScrolledWindow()
        scroll.set_vexpand(True)
        scroll.set_policy(Gtk.PolicyType.NEVER, Gtk.PolicyType.AUTOMATIC)

        outer = Gtk.Box(orientation=Gtk.Orientation.VERTICAL, spacing=12)
        outer.set_margin_top(12)
        outer.set_margin_bottom(12)

        # Back button row
        back_btn = Gtk.Button(label="← All Categories")
        back_btn.add_css_class("flat")
        back_btn.set_halign(Gtk.Align.START)
        back_btn.set_margin_start(12)
        back_btn.connect("clicked", lambda *_: self._nav_stack.set_visible_child_name("categories"))
        outer.append(back_btn)

        # "All <Category>" row at top
        list_box = Gtk.ListBox()
        list_box.set_selection_mode(Gtk.SelectionMode.NONE)
        list_box.add_css_class("boxed-list")
        list_box.set_margin_start(12)
        list_box.set_margin_end(12)

        all_row = Adw.ActionRow()
        all_row.set_title(f"All {label}")
        all_row.set_activatable(True)
        all_row.add_suffix(Gtk.Image.new_from_icon_name("go-next-symbolic"))
        all_row.connect("activated", lambda *_: self._show_apps(cat, f"All {label}"))
        list_box.append(all_row)

        if subcats:
            for sub_label, sub_cat in subcats:
                row = Adw.ActionRow()
                row.set_title(sub_label)
                row.set_activatable(True)
                row.add_suffix(Gtk.Image.new_from_icon_name("go-next-symbolic"))
                row.connect("activated",
                            lambda r, sl=sub_label, sc=sub_cat: self._show_apps(sc, sl))
                list_box.append(row)

        outer.append(list_box)
        scroll.set_child(outer)
        self._nav_stack.add_named(scroll, "subcategories")
        self._nav_stack.set_visible_child_name("subcategories")

    # ── App grid ──────────────────────────────────────────────────────────────

    def _show_apps(self, cat: str, label: str, back_to_home: bool = False):
        old = self._nav_stack.get_child_by_name("apps")
        if old:
            self._nav_stack.remove(old)

        outer = Gtk.Box(orientation=Gtk.Orientation.VERTICAL, spacing=0)
        outer.set_vexpand(True)

        # Back button — goes to Home when launched from Home, else to subcategories
        def _go_back(*_):
            if back_to_home:
                self._nav_stack.set_visible_child_name("categories")
                self._win._stack.set_visible_child_name("home")
            else:
                self._nav_stack.set_visible_child_name("subcategories")

        back_btn = Gtk.Button(label="← Home" if back_to_home else "← Categories")
        back_btn.add_css_class("flat")
        back_btn.set_halign(Gtk.Align.START)
        back_btn.set_margin_start(12)
        back_btn.set_margin_top(8)
        back_btn.connect("clicked", _go_back)
        outer.append(back_btn)

        scroll = Gtk.ScrolledWindow()
        scroll.set_vexpand(True)
        scroll.set_policy(Gtk.PolicyType.NEVER, Gtk.PolicyType.AUTOMATIC)

        self._apps_flow = Gtk.FlowBox()
        self._apps_flow.set_homogeneous(True)
        self._apps_flow.set_min_children_per_line(2)
        self._apps_flow.set_max_children_per_line(4)
        self._apps_flow.set_column_spacing(8)
        self._apps_flow.set_row_spacing(8)
        self._apps_flow.set_margin_start(12)
        self._apps_flow.set_margin_end(12)
        self._apps_flow.set_margin_top(8)
        self._apps_flow.set_margin_bottom(12)
        self._apps_flow.set_selection_mode(Gtk.SelectionMode.NONE)

        spinner = Gtk.Spinner()
        spinner.start()
        spinner.set_halign(Gtk.Align.CENTER)
        spinner.set_valign(Gtk.Align.CENTER)
        spinner.set_size_request(-1, 80)
        self._apps_flow.append(spinner)

        scroll.set_child(self._apps_flow)
        outer.append(scroll)
        self._nav_stack.add_named(outer, "apps")
        self._nav_stack.set_visible_child_name("apps")

        def _worker():
            try:
                result = _pkg.get_by_category(cat, limit=60)
                apps = result.get("items", []) if isinstance(result, dict) else result
            except Exception:
                apps = []
            GLib.idle_add(self._populate_apps, apps)

        threading.Thread(target=_worker, daemon=True).start()

    def _populate_apps(self, apps: list):
        # Clear spinner
        while True:
            child = self._apps_flow.get_first_child()
            if child is None:
                break
            self._apps_flow.remove(child)

        if not apps:
            lbl = Gtk.Label(label="No apps found")
            lbl.add_css_class("dim-label")
            self._apps_flow.append(lbl)
            return False

        for app in apps:
            card = self._make_app_card(app)
            self._apps_flow.append(card)
        return False

    def _make_app_card(self, app: dict) -> Gtk.Widget:
        btn = Gtk.Button()
        btn.add_css_class("card")
        btn.set_hexpand(True)

        box = Gtk.Box(orientation=Gtk.Orientation.HORIZONTAL, spacing=12)
        box.set_margin_top(10)
        box.set_margin_bottom(10)
        box.set_margin_start(12)
        box.set_margin_end(12)

        icon_img = Gtk.Image()
        icon_img.set_pixel_size(48)
        icon_path = resolve_icon_path(app)
        if icon_path and os.path.exists(icon_path):
            icon_img.set_from_file(icon_path)
        else:
            icon_img.set_from_icon_name(app.get("icon", "application-x-executable-symbolic"))

        text_box = Gtk.Box(orientation=Gtk.Orientation.VERTICAL, spacing=2,
                           valign=Gtk.Align.CENTER)
        name_lbl = Gtk.Label(label=app.get("name", ""))
        name_lbl.set_halign(Gtk.Align.START)
        name_lbl.add_css_class("body")
        name_lbl.set_ellipsize(Pango.EllipsizeMode.END)
        name_lbl.set_max_width_chars(20)

        sum_lbl = Gtk.Label(label=app.get("summary", ""))
        sum_lbl.set_halign(Gtk.Align.START)
        sum_lbl.add_css_class("caption")
        sum_lbl.add_css_class("dim-label")
        sum_lbl.set_ellipsize(Pango.EllipsizeMode.END)
        sum_lbl.set_max_width_chars(28)

        text_box.append(name_lbl)
        text_box.append(sum_lbl)
        box.append(icon_img)
        box.append(text_box)
        btn.set_child(box)

        btn.connect("clicked", lambda *_: self._win.navigate_to("detail", app=app))
        return btn

File: crates/ui-gtk/src/pages/explore.rs
// pages/explore.rs — Explore page: top-level category grid → sub-cat tiles + app grid

use gtk4::prelude::*;
use gtk4::{
    glib, Align, Box as GBox, Button, FlowBox, FlowBoxChild, Label, Orientation, ScrolledWindow,
    SelectionMode, Widget,
};
use libadwaita::prelude::*;
use libadwaita::{HeaderBar, NavigationPage, NavigationView, ToolbarView};
use std::cell::RefCell;
use std::rc::Rc;
use std::sync::{mpsc, Arc};
use std::time::Duration;

use rakuos_packages::NativeApp;
use super::icon_helper::load_app_icon;

// ── Category tree (mirrors Qt main.qml) ──────────────────────────────────────

struct TopCat {
    label: &'static str,
    cat: &'static str,
    icon: &'static str,
    subs: &'static [(&'static str, &'static str)],
}

const CATEGORIES: &[TopCat] = &[
    TopCat { label: "Games",        cat: "Game",        icon: "applications-games-symbolic", subs: &[
        ("Action",       "ActionGame"),
        ("Adventure",    "AdventureGame"),
        ("Arcade",       "ArcadeGame"),
        ("Board Games",  "BoardGame"),
        ("Card Games",   "CardGame"),
        ("Emulators",    "Emulator"),
        ("Kids",         "KidsGame"),
        ("Logic",        "LogicGame"),
        ("Role Playing", "RolePlaying"),
        ("Shooter",      "Shooter"),
        ("Simulation",   "Simulation"),
        ("Sports",       "SportsGame"),
        ("Strategy",     "StrategyGame"),
    ]},
    TopCat { label: "Internet",     cat: "Network",     icon: "network-wireless-symbolic", subs: &[
        ("Web Browsers",  "WebBrowser"),
        ("Email",         "Email"),
        ("Chat & IM",     "InstantMessaging"),
        ("File Sharing",  "FileTransfer"),
        ("News & RSS",    "News"),
        ("Remote Access", "RemoteAccess"),
    ]},
    TopCat { label: "Audio & Video", cat: "AudioVideo", icon: "audio-headphones-symbolic", subs: &[
        ("Music",   "Music"),
        ("Video",   "Video"),
        ("Editors", "AudioVideoEditing"),
        ("MIDI",    "Midi"),
        ("TV",      "TV"),
        ("Recorders","Recorder"),
        ("Mixers",  "Mixer"),
    ]},
    TopCat { label: "Graphics",     cat: "Graphics",    icon: "applications-graphics-symbolic", subs: &[
        ("2D Editors",  "2DGraphics"),
        ("3D Editors",  "3DGraphics"),
        ("Photography", "Photography"),
        ("Viewers",     "Viewer"),
        ("Publishing",  "Publishing"),
        ("Scanning",    "Scanning"),
    ]},
    TopCat { label: "Productivity", cat: "Office",      icon: "x-office-document-symbolic", subs: &[
        ("Office Suite", "WordProcessor"),
        ("Spreadsheets", "Spreadsheet"),
        ("Presentation", "Presentation"),
        ("Calendar",     "Calendar"),
        ("Finance",      "Finance"),
        ("Contacts",     "ContactManagement"),
        ("Project Mgmt", "ProjectManagement"),
    ]},
    TopCat { label: "Development",  cat: "Development", icon: "applications-development-symbolic", subs: &[
        ("IDEs",            "IDE"),
        ("Version Control", "RevisionControl"),
        ("Debuggers",       "Debugger"),
        ("Web Dev",         "WebDevelopment"),
        ("Database",        "Database"),
        ("Data Viz",        "DataVisualization"),
    ]},
    TopCat { label: "System",       cat: "System",      icon: "preferences-system-symbolic", subs: &[
        ("File Managers", "FileManager"),
        ("Terminal",      "TerminalEmulator"),
        ("Monitors",      "Monitor"),
        ("Security",      "Security"),
        ("Settings",      "Settings"),
        ("Package Mgmt",  "PackageManager"),
    ]},
    TopCat { label: "Utilities",    cat: "Utility",     icon: "applications-utilities-symbolic", subs: &[
        ("Archiving",   "Archiving"),
        ("Calculators", "Calculator"),
        ("Clocks",      "Clock"),
        ("Text Editors","TextEditor"),
        ("File Tools",  "FileTools"),
        ("Dictionary",  "Dictionary"),
    ]},
    TopCat { label: "Science",      cat: "Science",     icon: "applications-science-symbolic", subs: &[
        ("Astronomy",   "Astronomy"),
        ("Chemistry",   "Chemistry"),
        ("Electronics", "Electronics"),
        ("Engineering", "Engineering"),
        ("Geography",   "Geography"),
        ("Physics",     "Physics"),
    ]},
    TopCat { label: "Education",    cat: "Education",   icon: "applications-education-symbolic", subs: &[
        ("Languages",  "Languages"),
        ("Mathematics","Math"),
        ("Science",    "Science"),
        ("Literature", "Literature"),
        ("Geography",  "Geography"),
        ("Kids",       "KidsGame"),
    ]},
];

// ── Main explore page — only top-level category tiles ─────────────────────────

pub fn build(nav: Arc<NavigationView>) -> Widget {
    let scroll = ScrolledWindow::builder()
        .hscrollbar_policy(gtk4::PolicyType::Never)
        .vscrollbar_policy(gtk4::PolicyType::Automatic)
        .vexpand(true)
        .build();

    let main_box = GBox::builder()
        .orientation(Orientation::Vertical)
        .spacing(16)
        .margin_top(16)
        .margin_bottom(24)
        .margin_start(16)
        .margin_end(16)
        .build();
    scroll.set_child(Some(&main_box));

    let lbl = Label::builder()
        .label("Browse by Category")
        .halign(Align::Start)
        .css_classes(vec!["title-2".to_string()])
        .build();
    main_box.append(&lbl);

    let flow = FlowBox::builder()
        .selection_mode(SelectionMode::None)
        .min_children_per_line(2)
        .max_children_per_line(4)
        .column_spacing(12)
        .row_spacing(12)
        .homogeneous(true)
        .build();

    for top in CATEGORIES {
        let card = build_top_cat_card(top.label, top.cat, top.icon, top.subs, Arc::clone(&nav));
        let child = FlowBoxChild::new();
        child.set_child(Some(&card));
        child.set_focusable(false);
        flow.insert(&child, -1);
    }
    main_box.append(&flow);

    scroll.upcast()
}

// ── Top-level category card ───────────────────────────────────────────────────

fn build_top_cat_card(
    cat_label: &str,
    cat_id: &str,
    icon_name: &str,
    subs: &'static [(&'static str, &'static str)],
    nav: Arc<NavigationView>,
) -> Widget {
    let card = GBox::builder()
        .orientation(Orientation::Horizontal)
        .spacing(12)
        .css_classes(vec!["card".to_string()])
        .margin_top(4)
        .margin_bottom(4)
        .margin_start(4)
        .margin_end(4)
        .build();

    let icon = gtk4::Image::builder()
        .icon_name(icon_name)
        .pixel_size(32)
        .margin_start(12)
        .margin_top(12)
        .margin_bottom(12)
        .build();
    card.append(&icon);

    let name_lbl = Label::builder()
        .label(cat_label)
        .hexpand(true)
        .halign(Align::Start)
        .css_classes(vec!["body".to_string()])
        .build();
    card.append(&name_lbl);

    let chevron = gtk4::Image::builder()
        .icon_name("go-next-symbolic")
        .margin_end(12)
        .build();
    card.append(&chevron);

    let btn = Button::new();
    btn.set_child(Some(&card));
    btn.add_css_class("flat");

    let label_str = cat_label.to_string();
    let id_str    = cat_id.to_string();
    let nav_c = Arc::clone(&nav);

    btn.connect_clicked(move |_| {
        let page_w = build_category_page(&id_str, &label_str, subs, Arc::clone(&nav_c));
        let nav_page = NavigationPage::builder()
            .title(&label_str)
            .child(&page_w)
            .build();
        nav_c.push(&nav_page);
    });

    btn.upcast()
}

// ── Category page: sub-cat grid + top apps flow ───────────────────────────────

fn build_category_page(
    cat_id: &str,
    cat_label: &str,
    subs: &'static [(&'static str, &'static str)],
    nav: Arc<NavigationView>,
) -> Widget {
    let toolbar = ToolbarView::new();
    toolbar.add_top_bar(&HeaderBar::new());

    let scroll = ScrolledWindow::builder()
        .hscrollbar_policy(gtk4::PolicyType::Never)
        .vscrollbar_policy(gtk4::PolicyType::Automatic)
        .vexpand(true)
        .build();

    let main_box = GBox::builder()
        .orientation(Orientation::Vertical)
        .spacing(16)
        .margin_top(12)
        .margin_bottom(24)
        .margin_start(16)
        .margin_end(16)
        .build();
    scroll.set_child(Some(&main_box));

    // ── Sub-category tile grid ────────────────────────────────────────────────
    if !subs.is_empty() {
        let sub_lbl = Label::builder()
            .label("Sub-categories")
            .halign(Align::Start)
            .css_classes(vec!["title-3".to_string()])
            .build();
        main_box.append(&sub_lbl);

        let sub_flow = FlowBox::builder()
            .selection_mode(SelectionMode::None)
            .min_children_per_line(2)
            .max_children_per_line(3)
            .column_spacing(10)
            .row_spacing(10)
            .homogeneous(true)
            .build();

        for (sub_label, sub_cat) in subs {
            let tile = build_sub_cat_tile(sub_label, sub_cat, Arc::clone(&nav));
            let child = FlowBoxChild::new();
            child.set_child(Some(&tile));
            child.set_focusable(false);
            sub_flow.insert(&child, -1);
        }
        main_box.append(&sub_flow);
    }

    // ── Top apps header row ───────────────────────────────────────────────────
    let top_header = GBox::builder()
        .orientation(Orientation::Horizontal)
        .spacing(8)
        .build();

    let top_lbl = Label::builder()
        .label(&format!("Top {} Apps", cat_label))
        .halign(Align::Start)
        .hexpand(true)
        .css_classes(vec!["title-3".to_string()])
        .build();
    top_header.append(&top_lbl);

    let see_all_btn = Button::builder()
        .label("See all →")
        .css_classes(vec!["flat".to_string()])
        .visible(false)
        .build();
    top_header.append(&see_all_btn);
    main_box.append(&top_header);

    // ── App flow grid ─────────────────────────────────────────────────────────
    let app_flow = FlowBox::builder()
        .selection_mode(SelectionMode::None)
        .min_children_per_line(3)
        .max_children_per_line(8)
        .column_spacing(12)
        .row_spacing(12)
        .homogeneous(true)
        .build();
    main_box.append(&app_flow);

    // ── Load apps in background ───────────────────────────────────────────────
    let cat = cat_id.to_string();
    let (tx, rx) = mpsc::channel::<Vec<NativeApp>>();
    std::thread::spawn(move || {
        let apps = rakuos_packages::get_by_category(&cat).unwrap_or_default();
        let _ = tx.send(apps);
    });

    // Store all apps for "see all" toggle
    let stored_apps: Rc<RefCell<Vec<NativeApp>>> = Rc::new(RefCell::new(Vec::new()));
    let stored_c = Rc::clone(&stored_apps);
    let nav_c = Arc::clone(&nav);
    let flow_c = app_flow.clone();
    let btn_c = see_all_btn.clone();

    glib::timeout_add_local(Duration::from_millis(80), move || {
        match rx.try_recv() {
            Ok(apps) => {
                *stored_c.borrow_mut() = apps.clone();
                populate_app_flow(&flow_c, apps.iter().take(12), Arc::clone(&nav_c));
                if apps.len() > 12 {
                    btn_c.set_label(&format!("See all {} →", apps.len()));
                    btn_c.set_visible(true);
                }
                glib::ControlFlow::Break
            }
            Err(mpsc::TryRecvError::Empty) => glib::ControlFlow::Continue,
            Err(_) => glib::ControlFlow::Break,
        }
    });

    // "See all" reveals remaining apps
    let nav_sa = Arc::clone(&nav);
    see_all_btn.connect_clicked(move |btn| {
        let apps = stored_apps.borrow();
        populate_app_flow(&app_flow, apps.iter(), Arc::clone(&nav_sa));
        btn.set_visible(false);
    });

    toolbar.set_content(Some(&scroll));
    toolbar.upcast()
}

// ── Sub-category tile ─────────────────────────────────────────────────────────

fn build_sub_cat_tile(
    sub_label: &str,
    sub_cat: &str,
    nav: Arc<NavigationView>,
) -> Widget {
    let row = GBox::builder()
        .orientation(Orientation::Horizontal)
        .spacing(8)
        .css_classes(vec!["card".to_string()])
        .margin_top(2)
        .margin_bottom(2)
        .margin_start(2)
        .margin_end(2)
        .build();

    let lbl = Label::builder()
        .label(sub_label)
        .hexpand(true)
        .halign(Align::Start)
        .margin_start(12)
        .margin_top(14)
        .margin_bottom(14)
        .css_classes(vec!["body".to_string()])
        .build();
    row.append(&lbl);

    let chevron = gtk4::Image::builder()
        .icon_name("go-next-symbolic")
        .margin_end(10)
        .build();
    row.append(&chevron);

    let btn = Button::new();
    btn.set_child(Some(&row));
    btn.add_css_class("flat");

    let label_str = sub_label.to_string();
    let cat_str   = sub_cat.to_string();
    let nav_c = Arc::clone(&nav);

    btn.connect_clicked(move |_| {
        let page_w = build_sub_page(&cat_str, Arc::clone(&nav_c));
        let nav_page = NavigationPage::builder()
            .title(&label_str)
            .child(&page_w)
            .build();
        nav_c.push(&nav_page);
    });

    btn.upcast()
}

// ── Sub-category page: all apps in a flow grid ────────────────────────────────

fn build_sub_page(cat_id: &str, nav: Arc<NavigationView>) -> Widget {
    let toolbar = ToolbarView::new();
    toolbar.add_top_bar(&HeaderBar::new());

    let scroll = ScrolledWindow::builder()
        .hscrollbar_policy(gtk4::PolicyType::Never)
        .vscrollbar_policy(gtk4::PolicyType::Automatic)
        .vexpand(true)
        .build();

    let app_flow = FlowBox::builder()
        .selection_mode(SelectionMode::None)
        .min_children_per_line(3)
        .max_children_per_line(8)
        .column_spacing(12)
        .row_spacing(12)
        .homogeneous(true)
        .margin_top(12)
        .margin_bottom(24)
        .margin_start(16)
        .margin_end(16)
        .build();
    scroll.set_child(Some(&app_flow));

    let cat = cat_id.to_string();
    let (tx, rx) = mpsc::channel::<Vec<NativeApp>>();
    std::thread::spawn(move || {
        let apps = rakuos_packages::get_by_category(&cat).unwrap_or_default();
        let _ = tx.send(apps);
    });

    let nav_c = Arc::clone(&nav);
    glib::timeout_add_local(Duration::from_millis(80), move || {
        match rx.try_recv() {
            Ok(apps) => {
                if !apps.is_empty() {
                    populate_app_flow(&app_flow, apps.iter(), Arc::clone(&nav_c));
                }
                glib::ControlFlow::Break
            }
            Err(mpsc::TryRecvError::Empty) => glib::ControlFlow::Continue,
            Err(_) => glib::ControlFlow::Break,
        }
    });

    toolbar.set_content(Some(&scroll));
    toolbar.upcast()
}

// ── Helpers ───────────────────────────────────────────────────────────────────

fn populate_app_flow<'a>(
    flow: &FlowBox,
    apps: impl Iterator<Item = &'a NativeApp>,
    nav: Arc<NavigationView>,
) {
    while let Some(child) = flow.first_child() {
        flow.remove(&child);
    }
    for app in apps {
        let card = build_app_card(app, Arc::clone(&nav));
        let child = FlowBoxChild::new();
        child.set_child(Some(&card));
        child.set_focusable(false);
        flow.insert(&child, -1);
    }
}

fn build_app_card(app: &NativeApp, nav: Arc<NavigationView>) -> Widget {
    let card = GBox::builder()
        .orientation(Orientation::Vertical)
        .spacing(6)
        .width_request(120)
        .css_classes(vec!["card".to_string()])
        .margin_top(4)
        .margin_bottom(4)
        .build();

    let icon = load_app_icon(&app.icon_path, &app.icon_url, 56, &app.name);
    icon.set_halign(Align::Center);
    icon.set_margin_top(10);
    card.append(&icon);

    if app.installed {
        let badge = Label::builder()
            .label("✓")
            .css_classes(vec!["caption".to_string()])
            .halign(Align::End)
            .margin_end(8)
            .build();
        card.append(&badge);
    }

    let name_lbl = Label::builder()
        .label(&app.name)
        .halign(Align::Center)
        .wrap(false)
        .ellipsize(gtk4::pango::EllipsizeMode::End)
        .max_width_chars(14)
        .css_classes(vec!["caption-heading".to_string()])
        .margin_start(8)
        .margin_end(8)
        .margin_bottom(10)
        .build();
    card.append(&name_lbl);

    let btn = Button::new();
    btn.set_child(Some(&card));
    btn.add_css_class("flat");

    let (id, name, summary, ip, iu, src) = (
        app.id.clone(), app.name.clone(), app.summary.clone(),
        app.icon_path.clone(), app.icon_url.clone(), app.source.clone(),
    );
    btn.connect_clicked(move |_| {
        super::detail::push_detail(&nav, &id, &name, &summary, &ip, &iu, &src);
    });

    btn.upcast()
}

// build_list_row kept for any external callers
pub fn build_list_row(app: &NativeApp, nav: Arc<NavigationView>) -> Widget {
    build_app_card(app, nav)
}
